lib/src/rewrite.rs

```diff
@@ -379,34 +380,55 @@ impl<'settings, 'repo> DescendantRebaser<'settings, 'repo> {
         &self.rebased
     }
 
+    /// Panics if `parent_mapping` contains cycles
     fn new_parents(&self, old_ids: &[CommitId]) -> Vec<CommitId> {
-        // This should be a set, but performance of a vec is much better since we expect
-        // 99% of commits to have <= 2 parents.
-        let mut new_ids = vec![];
-        let mut add_parent = |id: &CommitId| {
-            // This can trigger if we abandon an empty commit, as both the empty commit and
-            // its parent are succeeded by the same commit.
-            if !new_ids.contains(id) {
-                new_ids.push(id.clone());
-            }
-        };
-        for old_id in old_ids {
-            if let Some(new_parent_ids) = self.parent_mapping.get(old_id) {
-                for new_parent_id in new_parent_ids {
-                    // The new parent may itself have been rebased earlier in the process
-                    if let Some(newer_parent_id) = self.rebased.get(new_parent_id) {
-                        add_parent(newer_parent_id);
-                    } else {
-                        add_parent(new_parent_id);
+        /// Update `new_ids`
+        fn single_substitution_round(
+            parent_mapping: &HashMap<CommitId, Vec<CommitId>>,
+            new_ids: &mut Vec<CommitId>,
+        ) -> bool {
+            let mut made_replacements = false;
+            // TODO(ilyagr): (Maybe?) optimize common case of replacements all
+            // being singletons. If CommitId-s were Copy. no allocations would be needed in
+            // that case, but it probably doesn't matter much while they are Vec<u8>-s.
+            let previously_new_ids = new_ids.split_off(0);
```

Comment:
nit: it's probably simpler to return new `new_ids` instead of faking in-place mutation.

Reply:
That's probably a good idea.

The in-place mutation would actually happen if I did the optimization I described in the comment or something like it. I keep going back and forth on whether it is a good idea, but am currently leaning against it.

I was also wondering of how to write the `while` loop then, but I guess I can do:

```rust
let mut new_ids = old_ids.into();
let mut made_replacements;
while { (new_ids, made_replacements) = single_substitution_round(...); made_replacements } {
   ...
}
```

Writing it out makes it look weird, though. I could also change it back to a `loop` loop. It seems to introduce a bit more visual noise, but is still OK. (Originally, I had the whole `single_substitution_round` function inlined in a `loop` loop, but I'll probably keep the function).

Reply:
Maybe it can be `while made_replacements { made_replacements = ... }`?

If you need in-lace mutation, `mem::take(&mut new_ids)` is probably better than `.split_off(0)`.

Reply:
How is this?

If you prefer `while made_replacements { made_replacements = ... }`, I can do this too. It felt a bit weird to start that variable out as `true`, though I could rename it to `need_another_round` or something.

Reply:
Looks better, thanks.

FWIW, if we want to optimize it for the common one-time one-element to one-element substitution, it might be better to flip the outer and inner loop.
```rust
while i < ids.len() {
    while let Some(replacements) = parent_mapping.get(&ids[i]) {
        // do some sanity checking here
        ids.splice(i..i + 1, replacements.iter().cloned());
    }
    i += 1;
}
```

Reply:
This is very nice, thank you! I'll put it in a separate PR. Here's a draft: https://github.com/ilyagr/jj/commit/refactor2.

My only very minor complaint is that if `replacements` is allowed to be empty, the necessary change to the logic becomes non-trivial. Maybe I should combine this with making `new_parents` handle that case (which would involve renaming it and writing a few tests).

Reply:
yeah, in that case we'll have to use labeled break/continue to reset the cycle counter for the new `i`-th item.

fwiw, at your option, cycle detection could (ab)use range iterator.
```rust
let mut iterations = 0..parent_mapping.len();
{
    iterations.next().expect("cycle detected");
}
```

Reply:
> yeah, in that case we'll have to use labeled break/continue to reset the cycle counter for the new i-th item.

Good point. I was thinking `i -= 1; break`, but I just realized that `i` is unsigned. :(

>  cycle detection could (ab)use range iterator

Cute. I'll probably use this tomorrow, unless I decide it's *too* cute.